File: java/src/main/java/problems/Problem008MergeTwoSortedLists.java
package problems;


public class Problem008MergeTwoSortedLists {
    public static class ListNode{int val; ListNode next; ListNode(int v){val=v;}}
    public static ListNode merge(ListNode a,ListNode b){
        ListNode d=new ListNode(0), t=d;
        while(a!=null&&b!=null){
            if(a.val<b.val){t.next=a;a=a.next;} else {t.next=b;b=b.next;}
            t=t.next;
        } t.next=(a!=null?a:b); return d.next;
    }
}

File: scala/src/main/scala/problems/Problem008MergeTwoSortedLists.scala

package problems

object Problem008MergeTwoSortedLists {
  final class ListNode(var x:Int, var next: ListNode)
  object ListNode { def apply(x:Int): ListNode = new ListNode(x, null) }

  def merge(a: ListNode, b: ListNode): ListNode = {
    val dummy = ListNode(0)
    var t = dummy; var p=a; var q=b
    while (p!=null && q!=null) {
      if (p.x < q.x) { t.next = p; p = p.next }
      else { t.next = q; q = q.next }
      t = t.next
    }
    t.next = if (p!=null) p else q
    dummy.next
  }
}

